### PR TITLE
Port yuzu-emu/yuzu#3173: "common: SPSCQueue: Notify after incrementing queue size."

### PR DESCRIPTION
**Pull request by FearlessTobi**
See yuzu-emu/yuzu#3173 for more details.

**Original description**:
Fixes a race condition.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/5062)

### CODE REVIEW

File: src/common/threadsafe_queue.h [hunk 1]
```diff
@@ -46,8 +46,15 @@ class SPSCQueue {
         ElementPtr* new_ptr = new ElementPtr();
         write_ptr->next.store(new_ptr, std::memory_order_release);
         write_ptr = new_ptr;
-        ++size;
 
+        const size_t previous_size{size++};
+
+        // Acquire the mutex and then immediately release it as a fence.
+        // TODO(bunnei): This can be replaced with C++20 waitable atomics when properly supported.
```
Comment (B3n30) [inline src/common/threadsafe_queue.h]: Nit: I don't think we will be able to enable c++20 in Citra because we would loose support for older Macs

Comment (FearlessTobi) [reply src/common/threadsafe_queue.h]: I don't think we should remove this comment though, in say 3 years the situation might be very different and then such a comment can be helpful.

**Comment by wwylele** [comment]
Can someone update the situation of #5012 after this merge?

**Comment by CookiePLMonster** [comment]
> Can someone update the situation of #5012 (is it fixed? is it still there but the code in the explanation needs an update?) after this merge?

I expect the issue in question to be fully addressed by this PR.

**Comment by FearlessTobi** [comment]
Can we merge this then? I guess #5012 could be closed afterwards?

**Comment by zhaowenlan1779** [comment]
I think #5012 is fixed with this.

**Pull request merged**